sync/sync.go

```diff
@@ -192,18 +201,70 @@ func (s *Synchronizer) syncBlocks(syncCtx context.Context) {
 				streamCancel()
 				fetchers.Wait()
 				verifiers.Wait()
+				pendingPollTicker.Stop()
+				pendingPollSem <- struct{}{}
 				return
 			default:
 				streamCtx, streamCancel = context.WithCancel(syncCtx)
 				nextHeight = s.nextHeight()
 				s.log.Warnw("Rolling back sync process", "height", nextHeight)
 			}
-		default:
+		case <-pendingPollTicker.C:
```

Comment:
>Could the pending block downloader be split into a separate Service that is run from

I thought of doing that too, but I don't think it warrant a new service. IMO, It can be thought of a part of sync process

Reply:
>  I don't think it warrants a new service

I am not fully convinced here. Can you share some of your reasoning? I'm mainly concerned about the concurrency complexity in the synchronizer, which a separate service should alleviate.

Kind of related: do we need to delay downloading the pending block until after we have reached the head? That seems to be a source of complexity.

Reply:
> I don't think we need fectcherSem (both max go routine and channel with the same buffer size seem redundant)

the problem is `stream.Go` is blocking, and could mess up the polling interval. Now that we poll pending in a separate go routine, that is no longer an issue.

Reply:
> the messages in the ticker channel keep building up and we end up bombarding the gateway with request for the pending block.

We should be able to consume extra ticks before calling `fetchAndStorePending`.

Reply:
Hmm, is reading len(channel) amount of messages is enough? Sounds like it could be racey to me. Semaphore helps with graceful shutdown as well. So I am still voting in favor of it:)